src/sentry/static/sentry/app/components/forms/genericField.tsx

```diff
@@ -0,0 +1,127 @@
+import PasswordField from 'app/components/forms/passwordField';
+import RangeField from 'app/components/forms/rangeField';
+import SelectAsyncField from 'app/components/forms/selectAsyncField';
+import SelectField from 'app/components/forms/selectField';
+import SelectCreatableField from 'app/components/forms/selectCreatableField';
+import TextField from 'app/components/forms/textField';
+import TextareaField from 'app/components/forms/textareaField';
+import FormState from 'app/components/forms/state';
+
+type FieldType =
+  | 'secret'
+  | 'range'
+  | 'bool'
+  | 'email'
+  | 'string'
+  | 'text'
+  | 'url'
+  | 'number'
+  | 'textarea'
+  | 'choice'
+  | 'select';
+
+type Config = {
+  required?: boolean;
+  help?: string;
+  name: string;
+  label?: string;
+  placeholder: string;
+  default: string;
+  readonly: boolean;
+  type: FieldType;
+  has_autocomplete: boolean;
+  choices: Array<[number | string, number | string]>;
+  url?: string; // TODO(ts): Required for SelectAsyncField
+};
+
+interface FormData {
+  [name: string]: string;
+}
+
+type Props = {
+  config: Config;
+  formData: FormData;
+  formErrors?: object;
+  formState: typeof FormState[keyof typeof FormState];
+  onChange: FormField['props']['onChange'];
+};
+
+const GenericField = ({
+  config,
+  formData = {},
+  formErrors = {},
+  formState,
+  onChange,
+}: Props) => {
+  const required = defined(config.required) ? config.required : true;
+  const fieldProps = {
+    ...config,
+    value: formData[config.name],
+    onChange,
+    label: config.label + (required ? '*' : ''),
+    placeholder: config.placeholder,
+    required,
+    name: config.name,
+    error: (formErrors || {})[config.name],
+    defaultValue: config.default,
+    disabled: config.readonly,
+    key: config.name,
+    formState,
+    help:
+      defined(config.help) && config.help !== '' ? (
+        <span dangerouslySetInnerHTML={{__html: config.help}} />
+      ) : null,
+  };
+
+  switch (config.type) {
+    case 'secret':
+      return <PasswordField {...fieldProps} />;
+    case 'range':
+      return <RangeField {...fieldProps} />;
+    case 'bool':
+      return <BooleanField {...fieldProps} />;
+    case 'email':
+      return <EmailField {...fieldProps} />;
+    case 'string':
+    case 'text':
+    case 'url':
+      if (fieldProps.choices) {
+        return <SelectCreatableField deprecatedSelectControl {...fieldProps} />;
+      }
+      return <TextField {...fieldProps} />;
+    case 'number':
+      return <NumberField {...fieldProps} />;
+    case 'textarea':
+      return <TextareaField {...fieldProps} />;
+    case 'choice':
+    case 'select':
+      // the chrome required tip winds up in weird places
+      // for select elements, so just make it look like
+      // it's required (with *) and rely on server validation
+      delete fieldProps.required;
+      if (fieldProps.has_autocomplete) {
+        return <SelectAsyncField deprecatedSelectControl url="" {...fieldProps} />;
```

Comment:
Where does the URL come from if this prop is blank?

Reply:
I was trying to work around making it required for all fields for genericField here and made it optional in `config` by just adding url for now (see: https://github.com/getsentry/sentry/pull/20662/files/4d64c75a065d3b478d5d9299fa1bbb673dbd9a98#diff-bac1125bad8e2ed91c8e1f48641d80acR42). I wasn't sure about narrowing it otherwise, aside from making config for `SelectAsyncField` a special case, unioning with the existing type, then narrowing right here based on `has_autocomplete`